This is a combination of 21 commits editing testing_controllers.md

jmarca opened the pull request:
update through running the first test

* replaced "phoenix.gen.json" with "phx.gen.json", updating to use
  context first

* fixed description accordingly

* in setup, gave instructions for creating a blank project.  Called
  that project hello_phoenix (in line with the original guide) rather
  than hello.

* instead of using phx.gen.schema, using phx.gen.context, which to me
  as a newcomer to phoenix seems like more 1.3 style.  Context is
  called Accounts, like other guides use.

* added a password field to the db table, plus made the email line
  unique, in the spirit of throwing a few things in there for interest
  to newbies like me.

* added a note that one would usually take a look at the generated
  migration, not just blindly apply it.

* Added in file names as a comment at the top of every elixir code
  block.  I found as a new user that was incredibly helpful, so I've
  tried to keep them.

* updated the code to use HelloPhoenix... rather than Hello.  I'm
  running all this in parallel, but I also already did it with the
  older name, so...sorry about that.

* corrected the mix test command.  The old one wouldn't have worked
  anyway, but also changed to `test/hello_phoneix_web/...

switch to project called hello, not hello_phoenix

it seems the tests intro section recommends making a new project
called hello

rerunning phx.gen.context to verify output

reran ecto.migrate to verify output

Added a little not about making sure tests pass

Before doing TDD, it is nice to run the existing tests and make sure
they pass.  From experience with Phoenix, I often forget to tweak
config/test.exs (generally I need the correct DB port).

wordsmithing setup output stuff

editing

patching the router.ex with users resources

Not sure about this one.  The current guide says to put the `resources
"/users"` inside the api pipe.  But the routing guide seems to
suggest that `resources "/users" ...` belongs in the `scope /api` block
instead.  So I moved it there and mentioned uncommenting that block.

Also, use the `except` option (something new here) because we don't
need to get forms generated for a JSON API.  Noted as such in text

fixed user_controller.ex

corrected Hello to HelloWeb

using Accounts.list_users() function  rather than Repo.all(User),
which to my eye looks more pre-1.3.

fix up user view

correct module name in user view to HelloWeb.UserView

minor changes:

* passing data around as "data", not "user"
* add a comment that we're not showing insert time, update time, and
  password, so new users understand that we *could* show those fields
  if we wanted to

added a note about using render_many

I'm still not super sold on why render_many exists rather than just
letting users call Enum.map.  Seems like language bloat to me, but I'm
sure it was argued back and forth at some point.

Main reason for adding the note is that I was doing it with Enum.map
directly and then came across render_many in the old guide, and had to
figure out why I'd want to use it rather than a map function.

change wording of first show test description

editing show test

added a section break, title

reworded first show test description

changed line to 41, because that's where I am now...

added file name and approx line number to code listing

changed test to use output from `setup/1` function, loading `user1`
from the context.

Also, namespaced the missing "show" action description to reinforce
that the modules map to path names.

added a note that test function is only extracting user1

show function in usercontroller

as before,
  * fixed the namespace from Hello to HelloWeb
  * using Accounts functions, not Repo

Added a note that Accounts.get_user! will throw if user not found.

fix spacing.

Update HelloWeb.UserView with show.json path

as before, fix namespaceing from Hello to HelloWeb.  Add the
"show.json" block.

update for "show with missing id" test

Added section header

noted that -1 is a good safe non-existent ID for the database.

fixed test to use conn from setup, and look for generic "resource not
found" rather than "user not found"

Added controller  handler for error

Talk about using put_status from Plug

talk about using ErrorView to gather up error handlers

add 400.json handler.

added a bigger finish

Since I did it, I added the rest of the test file, with a note that it
is fleshed out but there are lots of tests that still need to be
written and run.

fix capitalization of headings

and a final note about comparing to the JSON generator

fix links, some backticks, ugliness

fired up the mix docs.watch, fixed errors I found

jmarca commented:
Its a big edit.  No worries if you want to reject the whole and accept pieces.  I just squashed down 21 commits, but can also submit just the more important fixes.

Gazler commented on guides/docs/testing/testing_controllers.md:
```diff
@@ -133,169 +222,489 @@ defmodule HelloWeb.Router do
 
-  response = build_conn()
-  |> get(user_path(build_conn(), :show, user.id))
+test "Responds with user info if the user is found", %{conn: conn, user1: user} do
```
To create the user here, we should use:

    setup [:create_user]

Inside the describe and create a private `create_user` function. This is how the generators work for `gen.html` and `gen.json` https://github.com/phoenixframework/phoenix/blob/ede64f542e201934b36e6c8c5c9e8bd7046242ee/priv/templates/phx.gen.json/controller_test.exs

We should also mention how describe works. I would say something llike:

 > We use the `setup` macro in the `describe` to call a function before running each test. In this case, we want to create a user that we will display for the show action. The functions called by `setup` take an [ExUnit context](https://hexdocs.pm/ex_unit/ExUnit.Callbacks.html#module-context) (not to be confused with the contexts we are describing throughout this guide) and allow us to add additional fields when we return.

jmarca replied:
Sure.  Again, I just figured out how all this transaction wrapping stuff works today.

jmarca replied:
addressed...

Gazler commented on guides/docs/testing/testing_controllers.md:
````diff
@@ -133,169 +222,489 @@ defmodule HelloWeb.Router do
+When we run our test tells us we need a `HelloWeb.UserController.show/2` action.
 
 ```elixir
+# lib/hello_web/controllers/user_controller.ex
````
✂️

Gazler commented on guides/docs/testing/testing_controllers.md:
```diff
@@ -133,169 +222,489 @@ defmodule HelloWeb.Router do
-    users = Repo.all(User)
-    render conn, "index.json", users: users
+    users = Accounts.list_users()
+    render(conn, "index.json", data: users)
```
I don't think we should rename the `users` key to `data`. `users` is far more descriptive.(

jmarca replied:
Agreed.  Not sure why I started doing that.  I think at some point I was working towards getting "data" in the output JSON, and for some reason it crept upstream.

jmarca replied:
fixed

Gazler commented on guides/docs/testing/testing_controllers.md:
```diff
@@ -133,169 +222,489 @@ defmodule HelloWeb.Router do
 
-  def render("user.json", %{user: user}) do
-    %{name: user.name, email: user.email}
+  def render("user.json", %{data: user}) do
```
I think this was fine as :user

jmarca replied:
fixed

Gazler commented on guides/docs/testing/testing_controllers.md:
```diff
@@ -133,169 +222,489 @@ defmodule HelloWeb.Router do
-  def render("index.json", %{users: users}) do
-    %{data: render_many(users, HelloWeb.UserView, "user.json")}
+  def render("index.json", %{data: users}) do
+    %{data:
```
The formatting for the old version was fine, we can just add the `"show.html"` clause here.

jmarca replied:
fixed

Gazler commented on guides/docs/testing/testing_controllers.md:
```diff
@@ -133,169 +222,489 @@ defmodule HelloWeb.Router do
+
 When we run the test again, it passes.
 
+### No show
```
### Show when the user is not found

jmarca replied:
fixed

Gazler commented on guides/docs/testing/testing_controllers.md:
```diff
@@ -133,169 +222,489 @@ defmodule HelloWeb.Router do
 We want a HTTP code of 404 to notify the requester that this resource was not found, as well as an accompanying error message.
 
-Our controller action:
+Our controller action now needs to handle the error thrown by Ecto:
```
I don't think we should do this. The previous version should work without explicitly using a try/catch. If you wanted to match on the error, I'd recommend not using `Accounts.get_user!`, but defining an `Accounts.get_user` instead.

jmarca replied:
I don't really care much, but my thinking was the opposite of yours!

I didn't want to get into defining Accounts.get_user (something lots of guides do).  I chose instead to discuss catching exceptions, which I haven't seen much of in my reading of Phoenix guides. 

Thinking more now, I would argue (again, weakly, not my project I'll defer to your judgement) to keep the exception handling in.  The code generated by the phx.gen.json generator looks like this:

```
  def show(conn, %{"id" => id}) do
    user = Accounts.get_user!(id)
    render(conn, "show.json", user: user)
  end
```

If one passes an invalid id to that, it will throw.  The auto-generated tests do NOT test that case (I just double checked).  

I'm not sure why the code generators were designed with the get_THING!(id) semantics that advertise that they will throw a not found error, but if it throws then it has to be handled properly.  If everybody resorts to re-implementing get_THING! into get_THING (that does not throw), then why is the throwing version in the automatic generator in the first place?

Gazler replied:
When there is an `Ecto.NoResultsFound`, the `Plug.Exception` protocol is used to render the error. https://github.com/phoenixframework/phoenix_ecto/blob/master/lib/phoenix_ecto/plug.ex

I think the discussion around this worth opening another issue for though. So let's leave it for now and open a new issue.

jmarca replied:
fixed.

Gazler commented on guides/docs/testing/testing_controllers.md:
```diff
@@ -133,169 +222,489 @@ defmodule HelloWeb.Router do
 With those implemented, our tests pass.
 
-The rest of the controller is left to you to implement as practice. Happy testing!
+### Two down, six (or more) to go
```
I don't think we nee d a new heading here. We can say:

 > The rest of the controller is left for you to implement as practice. If you are not sure where to begin, it is worth using the Phoenix JSON generator and seeing what tests are automatically generated for you.

jmarca replied:
Agreed.  I interpret your intent to mean I should delete the sample test block as well, correct?  No one is really going to do it this way, they're going to use the phx.gen.json generator and edit that.

Gazler replied:
Your interpretation is correct.

jmarca replied:
fixed

jmarca commented on guides/docs/testing/testing_controllers.md:
````diff
@@ -5,49 +5,105 @@ We're going to take a look at how we might test drive a controller which has end
 What we are going for is a controller with the standard CRUD actions. We'll start with our test since we're TDDing this. Create a `user_controller_test.exs` file in `test/hello_web/controllers`
 
 ```elixir
+# test/hello_web/controllers/user_controller_test.exs
````
I would argue to leave the filename comments in, but only weakly (I don't really care).  It was half on, half off in the original Guite.  I made them all "on" in my edit, but then the intervening edit turned the filename comments all off.  

My "weak" support for keeping the file location comment line in is that the locations of the files changed when Phoenix was bumped to 1.3, and so it was a good way to notice that the old Guide was out of date. Also, as a newcomer to the project, it is sometimes hard to figure out where everything lives.

jmarca replied:
I've removed the filename comments

jmarca commented on guides/docs/testing/testing_controllers.md:
```diff
@@ -78,46 +134,79 @@ Create, show and update have more typical ways to fail because they need a way t
-  test "index/2 responds with all Users" do
-    users = [ User.changeset(%User{}, %{name: "John", email: "john@example.com"}),
-              User.changeset(%User{}, %{name: "Jane", email: "jane@example.com"}) ]
+  @user1_attrs %{email: "grobblefruit@example.org", name: "John", password: "surf and skate"}
```
I was basing off what the automatic tests look like.  When you run phx.gen.json, the automatic tests have attributes as in my version.   This then allows testing by comparing with variables, rather than with hard-coded strings.  

Not my project, so I'll do what you recommend, but hard-coded strings are bad.

jmarca commented on guides/docs/testing/testing_controllers.md:
```diff
@@ -78,46 +134,79 @@ Create, show and update have more typical ways to fail because they need a way t
+  @user2_attrs %{email: "varibiggles@example.org",  name: "Jane",  password: "coffee and beer"}
+
+  # setup creates users for all tests, and generates the conn
+  setup do
```
Yeah.  I didn't really understand what was going on when I did this.  I've been working on a related blog post this weekend, and now I get the purpose of the conn stuff, how transactions are implemented in the tests, and so on.

jmarca commented on guides/docs/testing/testing_controllers.md:
```diff
@@ -78,46 +134,79 @@ Create, show and update have more typical ways to fail because they need a way t
 
-    response = build_conn()
-    |> get(user_path(build_conn(), :index))
+    response = conn
```
I would instead suggest just insert one user, not two, and stick it in a test-local setup block, as is done in the auto-generated tests.

Gazler replied:
I agree!

jmarca replied:
Okay, I think what will work best here is to revert to the original for the first test, then in the second test case (show) refactor the inline creation to use the local setup call, fixture, private function as in the auto-generated tests.

jmarca replied:
fixed.  new test looks like

```
defmodule HelloWeb.UserControllerTest do
  use HelloWeb.ConnCase

  alias Hello.Accounts

  # setup generates the database connection
  setup do
    conn = build_conn()
    {:ok, conn: conn}
  end

  test "index/2 responds with all Users", %{conn: conn} do

    users = [%{name: "John", email: "john@example.com", password: "john pass"},
             %{name: "Jane", email: "jane@example.com", password: "jane pass"}]

    # create users local to this database connection and test
    [{:ok, user1},{:ok, user2}] = Enum.map(users, &Accounts.create_user(&1))

    response = conn
    |> get(user_path(conn, :index))
    |> json_response(200)

    expected = %{
      "data" => [
        %{ "name" => user1.name, "email" => user1.email },
        %{ "name" => user2.name, "email" => user2.email }
      ]
    }

    assert response == expected
  end
```

I will refactor this in the next section to use a local `setup` function to create users

jmarca commented on guides/docs/testing/testing_controllers.md:
```diff
@@ -133,169 +222,489 @@ defmodule HelloWeb.Router do
+and in fact `Enum.map` is called under the hood.  The main difference
+is that by using `render_many/4` instead of directly calling
+`Enum.map/2` is that the former benefits from library-quality error
+checking, properly handling missing values, and so on.
```
Okay.  This is the most clueless bit in my entire edit.  I truly have no idea why the render_one and render_many functions exist.  Seems like language bloat to me.  I originally was going to rip it out, but it shows up everywhere and in the autogenerated code, so there *must* be a reason for it, so I made something up...

jmarca commented on guides/docs/testing/testing_controllers.md:
```diff
@@ -133,169 +222,489 @@ defmodule HelloWeb.Router do
+    %{
+      name: user.name,
+      email: user.email
+      # skipping password, inserted_at, and updated_at
```
Okay.  I think I originally wrote that because I was also working the other way...editing the autogenerated stuff in which password is added but not needed.  And then (as a newcomer to Phoenix...really, I just started using it about two weeks ago) I'm not at all sure why the timestamps are stashed in the database, what purpose they would serve if they're never echoed back to the UI, etc.  

But sure, I'll delete the comment.

jmarca replied:
fixed

jmarca commented on guides/docs/testing/testing_controllers.md:
```diff
@@ -133,169 +222,489 @@ defmodule HelloWeb.Router do
+
 And with that, our test passes when we run it.
 
+### Time for the show
```
Punny headers are not on.  Got it!

jmarca replied:
fixed

jmarca commented:
On a quick read through I have zero *significant* objections to anything.  I made notes here and there where I disagree a bit with your changes, but please understand that I don't disagree *that* much...this isn't my project.

The only real irritant to me is the formatting.  Emacs automatically wraps lines, and my thumb automatically hits space twice.  (I'm old, and was taught two spaces after period back in high school when typewriters were a real thing for writing papers.)  Anyway, I was under the impression that markdown didn't care, that it would render the source content to fit the output.  Is that not the case?  I agree it looks ugly in the source file, though.  I'll do a search and replace to fix the line breaks and periods, but if you already have a canned way to do it that is more reliable please point it out. 

Ooo, never mind, I'm going to learn sed and write me a script.  Wordpress pukes on emacs-edited markdown, so it would be good to have something to filter out the autofill linebreaks.

Anyway I'll try to clean this up tomorrow.  That's the upside of being unemployed---I'm sure I'll have time!

And if you disagree with any or all of my push backs above, I'll pick your way.

Gazler commented:
Thanks for the feedback on my feedback!

 I was also taught double spacing, took a long time to train myself out of it! I didn't know pandoc could fix that, I should have guessed though, it can fix everything!

Let me know if I have missed anything or when this is ready for a final review.

jmarca commented on guides/docs/testing/testing_controllers.md:
````diff
@@ -5,49 +5,105 @@ We're going to take a look at how we might test drive a controller which has end
 ```
 
-Stuff is the context module name, Thing is the schema module name, things is the table name, and some_attr and another_attr are database columns on table things of type string. Don't run this command, we're going to explore test driving out a similar result to what a generator would give us.
+In this command, AllTheThings is the Context; Thing is the Schema;
````
fixed

jmarca commented on guides/docs/testing/testing_controllers.md:
```diff
@@ -5,49 +5,105 @@ We're going to take a look at how we might test drive a controller which has end
-Now run migrations:
+Ordinarily we would spend time tweaking the generated migration file
+(`priv/repo/migrations/<datetime>_create_users.exs`) to add things
+like non-null constraints and so on, but we don't care about that for
```
fixed

jmarca commented on guides/docs/testing/testing_controllers.md:
````diff
@@ -5,49 +5,105 @@ We're going to take a look at how we might test drive a controller which has end
 [info] == Migrated in 0.0s
 ```
 
+As a final check before we start developing, we run `mix test` and
````
fixed

jmarca commented on guides/docs/testing/testing_controllers.md:
````diff
@@ -5,49 +5,105 @@ We're going to take a look at how we might test drive a controller which has end
+```
+
+All of the tests should pass, but sometimes the database isn't
+specified properly in `config/test.exs`, or some other issue crops
````
fixed

jmarca commented on guides/docs/testing/testing_controllers.md:
````diff
@@ -133,169 +222,489 @@ defmodule HelloWeb.Router do
+We should get a new error now. Running the test informs us we don't have a `HelloWeb.UserController`. Let's add it, along with the `index/2` action we're testing. Our test description has us returning all users:
 
 ```elixir
+# lib/hello_web/controllers/user_controller.ex
````
fixed.  Wording is now: 

    Let's create that controller by opening the file `lib/hello_web/controllers/user_controller.ex` and adding the `index/2` action we're testing.

Could be word-smithed a bit better but good enough for now.

jmarca commented on guides/docs/testing/testing_controllers.md:
````diff
@@ -133,169 +222,489 @@ defmodule HelloWeb.Router do
 ```
 
-When we run the test again, our failing test tells us we have no view. Let's add it. Our test specifies a JSON format with a top key of `"data"`, containing an array of users with attributes `"name"` and `"email"`.
+When we run the test again, our failing test tells us module `HelloWeb.UserView` is not available. Let's add it. Our test specifies a JSON format with a top key of `"data"`, containing an array of users with attributes `"name"` and `"email"`.
````
fixed, using the wording:

    Let's add it by creating the file `lib/hello_web/views/user_view.ex`.

jmarca commented on guides/docs/testing/testing_controllers.md:
```diff
@@ -78,46 +134,79 @@ Create, show and update have more typical ways to fail because they need a way t
 When we run the test we get an error that we have no `user_path` function.
 
-In our router, we'll add a resource for `User` in our API pipe:
+In our router, we'll uncomment the `api` scope at the bottom of the
```
fixed

jmarca commented on guides/docs/testing/testing_controllers.md:
```diff
@@ -78,46 +134,79 @@ Create, show and update have more typical ways to fail because they need a way t
 
-Our expected data is a JSON response with a top level key of `"data"` containing an array of users that have `"name"` and `"email"` properties.
+Let's take a look at what's going on here. First, some attributes for
+two valid users are defined, and then a [`setup/1`
```
fixed

jmarca commented on guides/docs/testing/testing_controllers.md:
```diff
@@ -133,169 +222,489 @@ defmodule HelloWeb.Router do
 
-  def render("index.json", %{users: users}) do
-    %{data: render_many(users, HelloWeb.UserView, "user.json")}
+  def render("index.json", %{data: users}) do
```
fixed.  reverted to original view here

jmarca commented:
Okay, I'm pretty sure I addressed all your comments.  I left notes when I handled a comment except for the filename comments (I did those all at once)

Should I repackage this into a new pull request?

Gazler commented:
@jmarca No need to open a new PR. I just need to re-review this one.

Gazler commented:
Thanks for all the changes and thanks for the lengthy PR! <3
